File: wordsearchii.js
//Objective is the same as 'Word Search' except with multiple words this time

let board = 
[['o','a','a','n'],
 ['e','t','a','e'],
 ['i','h','k','r'],
 ['i','f','l','v']]
let words = ["oath","pea","eat","rain"]


//O(M(4⋅3^(L−1)) where M is the number of cells and L is the length of the longest word

let result = []
let root = buildTrie(words)

for (let i = 0; i < board.length; i++) {
    for (let j = 0; j < board[i].length; j++) {
        search(root, i, j)
    }
}

function buildTrie(words) {
    let root = {}
    
    for (let word of words) {
        //Reset back to the top of the root for each new word
        let node = root
        for (let char of word) {
            if (!node[char]) {
                node[char] = {}
            }
            node = node[char]
        }
        node.wordEnd = word
    }
    
    return root
}

function search(root, i, j) {
    if (root.wordEnd != null) {
        result.push(root.wordEnd)
        root.wordEnd = null
    }
    
    if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || root[board[i][j]] == null) {
        return
    }
    
    let temp = board[i][j]
    board[i][j] = '#'
    
    search(root[temp], i + 1, j)
    search(root[temp], i - 1, j)
    search(root[temp], i, j + 1)
    search(root[temp], i, j - 1)
    
    board[i][j] = temp
}

return result

File: wordsearchii.java
//Java Solution

class TrieNode {
    HashMap<Character, TrieNode> children;
    String word = null;
      
    public TrieNode() {
        children = new HashMap<Character, TrieNode>();
    }
  }
  
  class Solution {
      public List<String> findWords(char[][] board, String[] words) {
          TrieNode root = new TrieNode();
          for (String word : words) {
              TrieNode node = root;
              
              for (Character letter : word.toCharArray()) {
                  if (!node.children.containsKey(letter)) {
                      TrieNode newNode = new TrieNode();
                      node.children.put(letter, newNode);
                      node = newNode;
                  } else {
                      node = node.children.get(letter);
                  }
              }
              node.word = word;
          }
          
          List<String> result = new ArrayList<>();
          
          for (int i = 0; i < board.length; i++) {
              for (int j = 0; j < board[0].length; j++) {
                  dfs(i, j, root, board, result);
              }
          }
          
          return result;
      }
      
      public void dfs(int i, int j, TrieNode root, char[][] board, List result) {
          if (root.word != null) {
              result.add(root.word);
              root.word = null;
          }
          
          if (i < 0 || i >= board.length || j < 0 || j >= board[0].length || !root.children.containsKey(board[i][j])) {
              return;
          }
          
          char temp = board[i][j];
          board[i][j] = '#';
          
          dfs(i + 1, j, root.children.get(temp), board, result);
          dfs(i - 1, j, root.children.get(temp), board, result);
          dfs(i, j + 1, root.children.get(temp), board, result);
          dfs(i, j - 1, root.children.get(temp), board, result);
          
          board[i][j] = temp;
      }
  }